clippy_lints/src/mem_replace.rs

```diff
@@ -135,33 +135,66 @@ fn check_replace_option_with_none(cx: &LateContext<'_, '_>, src: &Expr<'_>, dest
     }
 }
 
-fn check_replace_with_uninit(cx: &LateContext<'_, '_>, src: &Expr<'_>, expr_span: Span) {
-    if let ExprKind::Call(ref repl_func, ref repl_args) = src.kind {
+fn check_replace_with_uninit(cx: &LateContext<'_, '_>, src: &Expr<'_>, dest: &Expr<'_>, expr_span: Span) {
+    if let ExprKind::MethodCall(method_path, _, method_args) = src.kind {
         if_chain! {
+            // check if replacement is `maybe_uninit::uninit().assume_init()`:
+            if method_path.ident.name == sym!(assume_init);
```

Comment:
If you want you could also make this function a diagnostic item in rustc and then reference it directly this way (instead of checking the name and doing a diagnostic item check for the `MaybeUninit` struct.

Reply:
Okay, I'm not very familiar with diagnostic items, but reading from the source code, it should be the API described in this issue right? https://github.com/rust-lang/rust-clippy/issues/5393

I will read more on this, are there any other references I should also read on diagnostic item API? Thanks :)

Reply:
I was sure I wrote docs, but I can't find anything. I apologize. I'll add some docs to the rustc guide.

TLDR: slap `#[rustc_diagnostic_item = "foobar")]` on a function or type declaration in rustc in order for `cx.tcx.is_diagnostic_item(sym::foobar, some_def_id)` to be true  if `some_def_id` is the `DefId` of said item. You also need to add `foobar` to the list in https://github.com/rust-lang/rust/blob/118b50524b79e565f017e08bce9b90a16c63634f/src/librustc_span/symbol.rs#L111

Reply:
So I checked the Rust source code, and found that `#[rustc_diagnostic_item = "assume_init"]` is already implemented: https://github.com/rust-lang/rust/blob/master/src/libcore/mem/maybe_uninit.rs#L495

I then implemented the diagnostic_item check and wow you're right, the code is greatly simplified and much more readable now. 

With this, I guess this implementation is pretty much done right? If yes, I can rebase everything into one and remove the draft status from this PR :+1:

Reply:
yes, this PR lgtm now! Thanks